docs/ProofOptimization/ProofOptimization.md

```diff
@@ -0,0 +1,597 @@
+
+<!-- %no-check -->
+```dafny
+lemma UnitIsUnique<T(!new)>(bop: (T, T) -> T, unit1: T, unit2: T)
+  requires IsUnital(bop, unit1)
+  requires IsUnital(bop, unit2)
+  ensures unit1 == unit2
+{
+  calc {
+    unit1;
+  == { assert IsRightUnital(bop, unit2); }
+    bop(unit1, unit2);
+  == { assert IsLeftUnital(bop, unit1); }
+    unit2;
+  }
+}
+
+```
+
+Here, the `calc` block establishes a chain of equalities, `unit1 == bop(unit1, unit2) == unit2` by using an intermediate assertion to justify each step. This is more information than Dafny strictly needs, but can make the proof both more readable and less variable.
+
+## Instantiating universally quantified formulas
+
+A universally quantified formula tells you that a certain fact is true for all values of a variable in a given domain. In many cases, however, you may know more about what specific instances need to be proved.
+
+
+Dafny can often bridge the gap between a universally quantified fact and an unquantified fact needed to make progress. However, it needs to do much less work if you can describe specific instantiations of interest.
+
+For example, an old commit in the `libraries` repository contains the following snippet:
+<!-- %no-check -->
+``` dafny
+lemma LemmaMulEqualityAuto()
+  ensures forall x: int, y: int, z: int {:trigger x * z, y * z } :: x == y ==> x * z == y * z
+
+...
+
+LemmaMulEqualityAuto();
+assert sum * pow == (z + cout * BASE()) * pow;
+```
+
+The assertion helps Dafny figure out how to instantiate the postcondition of `LemmaMulEqualityAuto`, making the proof more efficient.
+
+As we'll describe in more detail later, however, it can sometimes be better to avoid the quantifiers in the first place than to help Dafny instantiate them.
+
+# Hiding information
+
+There are several ways to structure a Dafny program to limit the number of facts that are available in scope at any particular point. First we'll go over a few that can be applied as small changes to existing programs, and then we'll cover some that are better applied early in the development process (though they can be applied to existing programs with more significant refactoring).
+
+## Subsumption
+
+The default behavior of an `assert` statement is to instruct the verifier to prove that a predicate is always true and then to assume that same predicate is true in later portions of the code, making it available to help prove other facts. This is called _subsumption_. On occasion, however, you may have an `assert` statement that exists to show a desired condition for its own sake, and that does not help prove later facts. In this case, you can disable subsumption, using `assert {:subsumption 0}`, to reduce the number of facts in scope in subsequent code.
```

Comment:
They are mentioned below (in the section "Opaque specification clauses", but perhaps it would make sense to move the "Subsumption" section after that one?